astropy/time/tests/test_methods.py

```diff
@@ -293,26 +251,96 @@ def test_broadcast(self, masked):
         assert np.may_share_memory(t2_broadcast.location, self.t2.location)
 
 
+class TestSetShape(ShapeSetup):
+    def test_shape_setting(self, masked):
+        # Shape-setting should be on the object itself, since copying removes
+        # zero-strides due to broadcasting.  Hence, this should be the only
+        # test in this class.
+        t0_reshape = self.t0.copy()
+        mjd = t0_reshape.mjd  # Creates a cache of the mjd attribute
+        t0_reshape.shape = (5, 2, 5)
+        assert t0_reshape.shape == (5, 2, 5)
+        assert mjd.shape != t0_reshape.mjd.shape  # Cache got cleared
+        assert np.all(t0_reshape.jd1 == self.t0._time.jd1.reshape(5, 2, 5))
+        assert np.all(t0_reshape.jd2 == self.t0._time.jd2.reshape(5, 2, 5))
+        assert t0_reshape.location is None
+        # But if the shape doesn't work, one should get an error.
+        t0_reshape_t = t0_reshape.T
+        with pytest.raises(AttributeError):
```

Comment:
Hmm, good question. It is actually a somewhat subtle bug, I now realize:
```
In [5]: np.ones((3, 2)).T.shape = 6,                                            
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-5-384542aa1cc8> in <module>
----> 1 np.ones((3, 2)).T.shape = 6,

AttributeError: incompatible shape for a non-contiguous array

In [6]: np.ones((3, 2)).T.shape = 1, 2                                          
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-6-570b08d46880> in <module>
----> 1 np.ones((3, 2)).T.shape = 1, 2

ValueError: cannot reshape array of size 6 into shape (1,2)
```
Since this bug was already present, I should solve it in a separate PR, so it can go into 4.0.x and 4.1.1. For now, I just made an issue, #10426